File: viewer-ui/src/lib/recovery.ts
/// What a pane's plugin last reported about getting it running again.
///
/// Pane metadata, not screen content: nothing here goes near an xterm instance,
/// so the terminal renderer is untouched by any of it.

/// The one `state` the server sends itself, meaning "nothing is pending for this
/// pane any more" — cancelled, expired, relaunched, or closed for good.
export const RECOVERY_CANCELLED = "cancelled";

export interface PaneRecovery {
  /// The plugin's own short label, e.g. `waiting_for_reset`. Uninterpreted.
  state: string;
  detail?: string;
  /// When the wait ends, in **unix epoch seconds**, absent when the plugin is not
  /// waiting on a clock.
  deadlineEpoch?: number;
  attempt: number;
}

/// The `recovery` control frame as the wire delivers it.
export interface RecoveryFrame {
  pane: number;
  state: string;
  detail?: string;
  deadline_epoch?: number;
  attempt: number;
}

export type RecoveryByPane = Record<number, PaneRecovery>;

/// Apply one `recovery` frame to the per-pane map.
///
/// A `cancelled` frame *removes* the pane instead of storing it: there is nothing
/// left to wait for, and keeping the label would leave a badge on a pane whose
/// recovery is over. The map is returned unchanged when there is nothing to
/// remove, so React skips a render for a frame that carries no news.
export function applyRecovery(
  current: RecoveryByPane,
  frame: RecoveryFrame,
): RecoveryByPane {
  if (frame.state === RECOVERY_CANCELLED) return forgetRecovery(current, frame.pane);
  return {
    ...current,
    [frame.pane]: {
      state: frame.state,
      detail: frame.detail,
      deadlineEpoch: frame.deadline_epoch,
      attempt: frame.attempt,
    },
  };
}

export function forgetRecovery(
  current: RecoveryByPane,
  pane: number,
): RecoveryByPane {
  if (!(pane in current)) return current;
  const next = { ...current };
  delete next[pane];
  return next;
}

/// Panes with a report that the page no longer lists — a pane whose process has
/// ended while its slot is held for a relaunch. It has no cell to put a badge in,
/// and it is exactly the one someone would want to release, so the panel's own
/// chrome shows it instead. Sorted so the row is stable across renders.
export function orphanRecovery(
  current: RecoveryByPane,
  panes: number[],
): number[] {
  return Object.keys(current)
    .map(Number)
    .filter((pane) => !panes.includes(pane))
    .sort((a, b) => a - b);
}

/// The deadline as the viewer's own local `HH:MM`, or `undefined` when there is
/// none.
///
/// `undefined` rather than a placeholder: a wrong wall-clock time reads as fact,
/// so the caller renders nothing instead. A non-finite or unrepresentable value is
/// treated the same way — it is a value no clock can place.
export function deadlineLabel(epochSeconds?: number): string | undefined {
  if (epochSeconds === undefined || !Number.isFinite(epochSeconds)) return undefined;
  const at = new Date(epochSeconds * 1000);
  if (Number.isNaN(at.getTime())) return undefined;
  return `${String(at.getHours()).padStart(2, "0")}:${String(
    at.getMinutes(),
  ).padStart(2, "0")}`;
}

/// The one-line summary shown in a pane's chrome: state, deadline, attempts.
///
/// The detail line is deliberately not folded in — it is shown as its own element
/// so a long one can be truncated without taking the state with it.
export function recoverySummary(report: PaneRecovery): string {
  const at = deadlineLabel(report.deadlineEpoch);
  const parts = [report.state];
  if (at) parts.push(`until ${at}`);
  if (report.attempt > 0) parts.push(`attempt ${report.attempt}`);
  return parts.join(" · ");
}

File: src/runtime/terminal/recovery.rs
//! What a pane's plugin last said about getting it running again.
//!
//! Held per pane and nowhere near the emulators: this is metadata the session
//! reports, not screen content, so it never touches a grid and a pane without a
//! report costs one absent map entry.

use crate::backend::PaneId;

use super::TerminalState;

/// The `state` a session sends once a pane's recovery is over without having
/// succeeded. Mirrors the hub's own constant; the wire is the contract between
/// them, so the value is asserted rather than shared.
pub const RECOVERY_CANCELLED: &str = "cancelled";

/// One pane's latest recovery report.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct PaneRecovery {
    /// The plugin's own short label, e.g. `waiting_for_reset`.
    pub state: String,
    /// A short human line, when the plugin gave one.
    pub detail: Option<String>,
    /// When the wait ends, in unix epoch seconds. `None` when the plugin is not
    /// waiting on a clock — the renderer then shows no time at all rather than
    /// inventing one.
    pub deadline_epoch: Option<i64>,
    pub attempt: u32,
}

impl TerminalState {
    /// Record what the session reports about `pane`.
    ///
    /// A `cancelled` report *clears* the entry instead of storing it: there is
    /// nothing left to wait for, and keeping the label would leave a badge on a
    /// pane whose recovery is over.
    pub(super) fn apply_recovery(
        &mut self,
        pane: PaneId,
        state: String,
        detail: Option<String>,
        deadline_epoch: Option<i64>,
        attempt: u32,
    ) {
        if state == RECOVERY_CANCELLED {
            self.recovery.remove(&pane);
            return;
        }
        self.recovery.insert(
            pane,
            PaneRecovery {
                state,
                detail,
                deadline_epoch,
                attempt,
            },
        );
    }

    /// What `pane`'s plugin last reported, if anything.
    pub fn recovery_for(&self, pane: PaneId) -> Option<&PaneRecovery> {
        self.recovery.get(&pane)
    }

    /// The one report a person is looking at, and the one the cancel key acts
    /// on: the focused pane's own report first, failing that a report for a
    /// pane this client no longer lists (its process ended while its slot is
    /// held for a relaunch). That pane cannot be focused, and it is exactly
    /// the one someone would want to release. Lowest id wins so the display
    /// and the key can never disagree about which one that is.
    pub fn recovery_focus(&self) -> Option<(PaneId, &PaneRecovery)> {
        if let Some(pane) = self.active_pane_id()
            && let Some(report) = self.recovery.get(&pane)
        {
            return Some((pane, report));
        }
        self.recovery
            .iter()
            .filter(|(pane, _)| !self.panes.iter().any(|p| p.id == **pane))
            .min_by_key(|(pane, _)| **pane)
            .map(|(pane, report)| (*pane, report))
    }

    /// Whether the cancel key would do anything. Single source for the key gate
    /// and the hint row, so a hint can never advertise a no-op.
    pub fn can_cancel_recovery(&self) -> bool {
        self.recovery_focus().is_some()
    }

    /// Ask the session to give up on the recovery a person is looking at.
    /// Nothing is cleared here: the entry goes when the session broadcasts
    /// `cancelled` (which tells every other client too) — assuming it locally
    /// would hide a cancellation the session refused.
    pub fn cancel_recovery(&mut self) {
        let Some((pane, _)) = self.recovery_focus() else {
            return;
        };
        if let Some(backend) = &mut self.backend {
            backend.cancel_recovery(pane);
        }
    }
}
